--- linux-4.4.0-chi440/drivers/i2c/busses/i2c-designware-platdrv.new.c	2016-01-10 16:01:32.000000000 -0700
+++ linux-4.4.0-chi440/drivers/i2c/busses/i2c-designware-platdrv.c	2016-01-21 21:11:53.016837107 -0700
@@ -107,6 +107,8 @@ static int dw_i2c_acpi_configure(struct
 	dw_i2c_acpi_params(pdev, "FMCN", &dev->fs_hcnt, &dev->fs_lcnt,
 			   &dev->sda_hold_time);
 
+	dev->sda_hold_time = 30;		/* (*) */
+
 	id = acpi_match_device(pdev->dev.driver->acpi_match_table, &pdev->dev);
 	if (id && id->driver_data)
 		dev->accessor_flags |= (u32)id->driver_data;
